internal/controller/utils/utils.go

```diff
@@ -0,0 +1,32 @@
+package utils
```

Comment:
I'd generally avoid `utils` packages. They tend to get abused. This could be an `errors` or `retry` package if you're looking for a name.

Reply:
Decomposed utils -> conditions and retry.